linera-client/src/chain_listener.rs

```diff
@@ -85,57 +85,68 @@ pub trait ClientContext: 'static {
-    config: ChainListenerConfig,
-    listening: Arc<Mutex<HashSet<ChainId>>>,
+pub struct ChainListener<C: ClientContext> {
+    context: Arc<Mutex<C>>,
+    storage: C::Storage,
+    config: Arc<ChainListenerConfig>,
+    listening: Arc<std::sync::Mutex<HashSet<ChainId>>>,
 }
 
-impl ChainListener {
+impl<C: ClientContext> Clone for ChainListener<C> {
+    fn clone(&self) -> Self {
+        Self {
+            context: self.context.clone(),
+            storage: self.storage.clone(),
+            config: self.config.clone(),
+            listening: self.listening.clone(),
+        }
+    }
+}
+
+/// A `ChainListener` together with a specific client.
+pub struct ChainClientListener<C: ClientContext> {
+    client: ContextChainClient<C>,
+    listener: ChainListener<C>,
+    timeout: Timestamp,
+}
+
+impl<C: ClientContext> ChainListener<C> {
     /// Creates a new chain listener given client chains.
-    pub fn new(config: ChainListenerConfig) -> Self {
+    pub fn new(config: ChainListenerConfig, context: Arc<Mutex<C>>, storage: C::Storage) -> Self {
         Self {
-            config,
+            storage,
+            context,
+            config: Arc::new(config),
             listening: Default::default(),
         }
     }
 
     /// Runs the chain listener.
-    pub async fn run<C>(self, context: Arc<Mutex<C>>, storage: C::Storage)
-    where
-        C: ClientContext,
-    {
+    pub async fn run(self) {
         let chain_ids = {
-            let guard = context.lock().await;
+            let guard = self.context.lock().await;
             let mut chain_ids = BTreeSet::from_iter(guard.wallet().chain_ids());
             chain_ids.insert(guard.wallet().genesis_admin_chain());
             chain_ids
         };
         for chain_id in chain_ids {
-            Self::run_with_chain_id(
-                chain_id,
-                context.clone(),
-                storage.clone(),
-                self.config.clone(),
-                self.listening.clone(),
-            );
+            self.run_with_chain_id(chain_id);
         }
     }
 
+    /// Spawns a task running the listener for the given chain, if it is not already running.
     #[instrument(level = "trace", skip_all, fields(?chain_id))]
-    fn run_with_chain_id<C>(
-        chain_id: ChainId,
-        context: Arc<Mutex<C>>,
-        storage: C::Storage,
-        config: ChainListenerConfig,
-        listening: Arc<Mutex<HashSet<ChainId>>>,
-    ) where
-        C: ClientContext,
-    {
+    fn run_with_chain_id(&self, chain_id: ChainId) {
+        if !self.listening.lock().unwrap().insert(chain_id) {
+            // If we are already listening to notifications, there's nothing to do.
+            // This can happen if we download a child before the parent
+            // chain, and then process the OpenChain message in the parent.
+            return;
+        }
+        let listener = self.clone();
         let _handle = linera_base::task::spawn(
             async move {
```

Comment:
Maybe this should be `in_current_span`?

Reply:
Ah, good catch! :+1:

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/3726/commits/b62d410eed39377d1847d40bf21c1b4a88d1747d.